### PR TITLE
Migrate www/* to the new data service

### PR DESCRIPTION
**Pull request by tothandras**
TODOs:
- [x] migrate base
- [x] fix log viewer
- [x] migrate waterfall-view
- [ ] migrate console-view

### CODE REVIEW

File: www/base/.editorconfig [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# editorconfig.org
```
Comment (sa2ajj) [inline www/base/.editorconfig]: do we really need this?

File: master/docs/developer/www.rst [hunk 2]
```diff
@@ -620,7 +620,7 @@ Several methods are added to each "restangularized" objects, aside from get(), p
 DataService
 .............
 
-DataService is the future replacement of BuildbotService for accessing the Buildbot data API.
+DataService is the replacement of BuildbotService for accessing the Buildbot data API.
 It has a modern interface for accessing data. It uses IndexedDB for storing cached data as a single data store,
```
Comment (sa2ajj) [inline master/docs/developer/www.rst]: While you are at it, could you please bring it up to the one sentence per line policy?

File: www/base/guanlecoja/config.coffee [hunk 3]
```diff
@@ -3,7 +3,7 @@
 #   This module contains all configuration for the build process
 #
 ### ###############################################################################################
-ANGULAR_TAG = "~1.3.1"
+ANGULAR_TAG = "~1.4.4"
```
Comment (sa2ajj) [inline www/base/guanlecoja/config.coffee]: Is it something forgotten in an earlier PR?

**Comment by sa2ajj** [comment]
Coffee script/jade is not my area; I shouldn't have started the review...

File: www/base/guanlecoja/config.coffee (hunk 3 above)
Comment (tothandras) [inline www/base/guanlecoja/config.coffee]: It's not yet merged, something was problematic with it: https://github.com/buildbot/buildbot/pull/1799

**Comment by sa2ajj** [comment]
We have some test failures:

```
[ERROR]
Traceback (most recent call last):
  File "/home/travis/build/buildbot/buildbot/pkg/test_buildbot_pkg.py", line 81, in test_develop
    self.run_setup("develop")
  File "/home/travis/build/buildbot/buildbot/pkg/test_buildbot_pkg.py", line 57, in run_setup
    check_call("python setup.py " + cmd, shell=True, cwd=self.path)
  File "/opt/python/2.7.9/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'python setup.py develop' returned non-zero exit status 1

test_buildbot_pkg.BuildbotWWWPkg.test_develop
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/travis/build/buildbot/buildbot/pkg/test_buildbot_pkg.py", line 85, in test_develop_via_pip
    check_call("pip install -e .", shell=True, cwd=self.path)
  File "/opt/python/2.7.9/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'pip install -e .' returned non-zero exit status 1

test_buildbot_pkg.BuildbotWWWPkg.test_develop_via_pip
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/travis/build/buildbot/buildbot/pkg/test_buildbot_pkg.py", line 75, in test_egg
    self.run_setup("bdist_egg")
  File "/home/travis/build/buildbot/buildbot/pkg/test_buildbot_pkg.py", line 57, in run_setup
    check_call("python setup.py " + cmd, shell=True, cwd=self.path)
  File "/opt/python/2.7.9/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'python setup.py bdist_egg' returned non-zero exit status 1

test_buildbot_pkg.BuildbotWWWPkg.test_egg
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/travis/build/buildbot/buildbot/pkg/test_buildbot_pkg.py", line 66, in test_install
    self.run_setup("install")
  File "/home/travis/build/buildbot/buildbot/pkg/test_buildbot_pkg.py", line 57, in run_setup
    check_call("python setup.py " + cmd, shell=True, cwd=self.path)
  File "/opt/python/2.7.9/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'python setup.py install' returned non-zero exit status 1

test_buildbot_pkg.BuildbotWWWPkg.test_install
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/travis/build/buildbot/buildbot/pkg/test_buildbot_pkg.py", line 89, in test_sdist
    self.run_setup("sdist")
  File "/home/travis/build/buildbot/buildbot/pkg/test_buildbot_pkg.py", line 57, in run_setup
    check_call("python setup.py " + cmd, shell=True, cwd=self.path)
  File "/opt/python/2.7.9/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'python setup.py sdist' returned non-zero exit status 1

test_buildbot_pkg.BuildbotWWWPkg.test_sdist
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/travis/build/buildbot/buildbot/pkg/test_buildbot_pkg.py", line 70, in test_wheel
    self.run_setup("bdist_wheel")
  File "/home/travis/build/buildbot/buildbot/pkg/test_buildbot_pkg.py", line 57, in run_setup
    check_call("python setup.py " + cmd, shell=True, cwd=self.path)
  File "/opt/python/2.7.9/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'python setup.py bdist_wheel' returned non-zero exit status 1

test_buildbot_pkg.BuildbotWWWPkg.test_wheel
```

File: master/docs/developer/www.rst (hunk 2 above)
Comment (tothandras) [inline master/docs/developer/www.rst]: will do

**Comment by tothandras** [comment]
I was not expecting passing tests as I removed buildbotService and some modules still depends on it (eg. waterfall, console plugins). I will look into this after I finish all the migration.

File: www/base/guanlecoja/config.coffee (hunk 3 above)
Comment (sa2ajj) [inline www/base/guanlecoja/config.coffee]: Is it a pre-requisite for this PR?

Comment (tothandras) [inline www/base/guanlecoja/config.coffee]: No, it's not. I can rebase if it will be merged before this.

Comment (sa2ajj) [inline www/base/guanlecoja/config.coffee]: Then, I think, I'd be the best to do so, unless @tardyp sees it differently

File: www/base/.editorconfig [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+end_of_line = lf
+charset = utf-8
+trim_trailing_whitespace = true
+insert_final_newline = true
+
+[*.{js,json}]
+indent_size = 2
+
+[*.md]
+trim_trailing_whitespace = false
```
Comment (tardyp) [inline www/base/.editorconfig]: why?

File: www/base/guanlecoja/config.coffee (hunk 3 above)
Comment (tardyp) [inline www/base/guanlecoja/config.coffee]: well, I am not against upgrading angular if that works.

File: www/base/guanlecoja/config.coffee [hunk 5]
```diff
@@ -53,6 +50,17 @@ config =
             "reconnectingWebsocket":
                 version: "master"
                 files: ["reconnecting-websocket.js"]
+            'buildbot-data':
+                version: '~1.0.12'
+                files: 'dist/scripts.js'
+            # TODO these are dependencies of buildbot-data, could be included
```
Comment (tardyp) [inline www/base/guanlecoja/config.coffee]: You can look in guanlecoja-ui how it generates a dist/vendors.js

File: www/base/src/app/builders/build/build.controller.coffee [hunk 6]
```diff
@@ -31,42 +32,55 @@ class Build extends Controller
...
-                    caption: "Builders"
-                    sref: "builders"
-                ,
-                    caption: builder.name
-                    sref: "builder({builder:#{builderid}})"
-                ,
-                    caption: build.number
-                    sref: "build({build:#{buildnumber}})"
-            ]
+        opened = dataService.open($scope)
```
Comment (tardyp) [inline www/base/src/app/builders/build/build.controller.coffee]: opened? sounds like a bad identifier

File: www/base/src/app/builders/step/step.controller.coffee [hunk 7]
```diff
@@ -1,20 +1,14 @@
...
+        buildnumber = dataUtilsService.numberOrString($stateParams.build)
+        stepnumber = dataUtilsService.numberOrString($stateParams.step)
+        opened.getBuilders(builderid).then (builders) ->
+            $scope.builder = builder = builders[0]
+            builder.getBuilds(buildnumber).then (builds) ->
+                $scope.build = build = builds[0]
+                build.getSteps(stepnumber).then (steps) ->
+                    step = steps[0]
+                    step.loadLogs()
+                    $scope.step = publicFieldsFilter(step)
```
Comment (tardyp) [inline www/base/src/app/builders/step/step.controller.coffee]: no breadcrumb configuration?

File: www/base/src/app/common/filters/publicFields.filter.coffee [hunk 8]
```diff
@@ -0,0 +1,6 @@
+class PublicFields extends Filter('common')
+    constructor: ->
+        return (object) ->
+            for k of object
+                if k.indexOf('_') == 0 then delete object[k]
```
Comment (tardyp) [inline www/base/src/app/common/filters/publicFields.filter.coffee]: delete?
Are you modifying the original object? sounds bad.
why dont you rather build a copy

File: www/base/src/app/common/services/buildbot/buildbot.service.coffee [hunk 9]
```diff
@@ -1,203 +0,0 @@
-class BuildbotService extends Factory
```
Comment (tardyp) [inline www/base/src/app/common/services/buildbot/buildbot.service.coffee]: yeah!

File: www/base/src/app/layout.jade [hunk 10]
```diff
@@ -14,17 +14,11 @@ html.no-js(xmlns:ng='http://angularjs.org', xmlns:app='ignored')
         block footer
 
     script(src="scripts.js?_#{(new Date()).getTime()}")
-    | {% for app in config.plugins -%}
+    | {% for app in config.plugins %}
```
Comment (tardyp) [inline www/base/src/app/layout.jade]: please see http://jinja.pocoo.org/docs/dev/templates/ why the -%} is important

File: www/base/.editorconfig (hunk 4 above)
Comment (tothandras) [inline www/base/.editorconfig]: I like it :) 
Shall I remove this?

File: www/base/guanlecoja/config.coffee (hunk 5 above)
Comment (tothandras) [inline www/base/guanlecoja/config.coffee]: I see, so `vendors_apart: true` and include `dist/vendors.js` here as an array.

File: www/base/src/app/builders/build/build.controller.coffee (hunk 6 above)
Comment (tothandras) [inline www/base/src/app/builders/build/build.controller.coffee]: With my limited English skills I couldn't find a better one. I am open to anything else.

File: www/base/src/app/common/filters/publicFields.filter.coffee (hunk 8 above)
Comment (tothandras) [inline www/base/src/app/common/filters/publicFields.filter.coffee]: well.. lets say I was tired :)

File: www/base/src/app/builders/build/build.controller.coffee (hunk 6 above)
Comment (tardyp) [inline www/base/src/app/builders/build/build.controller.coffee]: Just data..

**Comment by tothandras** [comment]
@tardyp sorry for the delay, I was having a hard time. This PR is already huge, I think this can be merged as soon as I fix the log viewer and I will migrate the console_view in another PR.

**Comment by tothandras** [comment]
@tardyp The log viewer seems to be working fine. I think it just fixed itself.

**Comment by shanzi** [comment]
:+1:

**Comment by tardyp** [comment]
travis fails.. because of angular version conflict. I'm looking at it

**Comment by tardyp** [comment]
This does not work for me.
Did you test it outside of the proxy?

Going in the builder page, nothing happens. no ws socket is created, no REST api is called

**Comment by tothandras** [comment]
Everything works fine for me, both with base and md_base. Try to delete the cache first: `indexedDB.deleteDatabase('bbCache')`. Something might be stuck from a previous version.

**Comment by tothandras** [comment]
@tardyp I've also applied your PR and updated the version of `buildbot-data`.

File: www/data_module/src/services/socket/socket.service.coffee [hunk 11]
```diff
@@ -71,15 +71,20 @@ class Socket extends Service
 
             getUrl: ->
                 host = $location.host()
-                port = if $location.port() is 80 then '' else ':' + $location.port()
-                return "ws://#{host}#{port}/ws"
+                protocol = if $location.protocol() is 'https' then 'wss' else 'ws'
+                defaultport = if $location.protocol() is 'https' then 443 else 80
+                path = $location.path()
```
Comment (tardyp) [inline www/data_module/src/services/socket/socket.service.coffee]: $location.path() return 'bar' when url is 'http://some.com/foo/#/bar
while we want /foo/. See last version of my PR

Comment (tothandras) [inline www/data_module/src/services/socket/socket.service.coffee]: Where is the last version? I've just merged the PR against buildbot-data.

**Comment by tardyp** [comment]
The latest version is here:
https://github.com/tothandras/buildbot-data/pull/2

I've finally managed to make this live on my own buildbot_travis instance.
I've seen some regressions though
- The performance seen is notably slower.
- Lots of links in the breadcrumb does not work (builder page, build page, log page)


I propose we log those two bugs, and followup in next PRs. luckily we released beta 3 yesterday, so this can wait beta4

however, you need to get travis work for this PR.

**Comment by tothandras** [comment]
@tardyp that is what I was talking about last time, did you noticed any difference between different browsers?

**Comment by tothandras** [comment]
@tardyp we should talk about this how to resolve, you can reach me on hangouts all day.

**Comment by tardyp** [comment]
@tothandras you need to update gulp as well..

**Pull request merged**